netmaster/k8snetwork/networkpolicy.go

```diff
@@ -391,8 +803,626 @@ func InitK8SServiceWatch(listenURL string, isLeader func() bool) error {
+	toPodIPs []string,
+	ingressRules []client.Rule,
+	isAdd bool) *[]client.Rule {
+
+	var err error
+	ruleList := ingressRules
+
+	//policyCtx := k8sNet.policyRules[np.PodSelector.PolicyName]
+
+	//Ingress Spec To section Pods
+	for _, toIps := range toPodIPs {
+		//npLog.Infof("ruleList:%v", ruleList)
+		//Rebuild Rule List to add To Ips
+		for _, rule := range ruleList {
+			//Update To src Ip section in Rule
+			rule.ToIpAddress = toIps
+			//Generate RuleID :XXX: Should look for better approach
+			rule.RuleID = k8sutils.PolicyToRuleIDUsingIps(
+				toIps, rule.FromIpAddress,
+				rule.Port, rule.Protocol,
+				np.PodSelector.PolicyName)
+
+			npLog.Infof("RulID:%v", rule.RuleID)
+			//Update Policy Name cache with policy Id
+			if isAdd {
+				if err = k8sNet.createRule(&rule); err != nil {
+					npLog.Errorf("failed: rules in-policy %+v, %v",
+						np.PodSelector, err)
+					return nil
+				}
+				//Update RuleID in cache Db
+				//XXX:Should use Hash Set instead of slice
+				// to aviod duplicate Ruleid insertion
+				//policyCtx = append(policyCtx, rule.RuleID)
+			} else { //Policy Delete
+				if err = k8sNet.deleteRule(rule.TenantName,
+					rule.PolicyName,
+					rule.RuleID); err != nil {
+					npLog.Errorf("failed: del in-policy %+v, %v",
+						np.PodSelector, err)
+					return nil
+				}
+				//	for idx, r := range policyCtx {
+				//		if r == rule.RuleID {
+				//			policyCtx = append(policyCtx[0:idx],
+				//				policyCtx[idx+1:]...)
+				//		}
+				//	}
+			}
+			policyMap := k8sNet.policyPerEpg[np.PodSelector.GroupName]
+			policy := policyMap[np.PodSelector.PolicyName]
+			if isAdd {
+				policy = append(policy, rule.RuleID)
+				npLog.Infof("RuleId:%v is added into Policy:%v",
+					rule.RuleID, np.PodSelector.PolicyName)
+			} else {
+				for idx, r := range policy {
+					if r == rule.RuleID {
+						policy = append(policy[0:idx],
+							policy[idx+1:]...)
+						npLog.Infof("RuleId :%v deleted ",
+							rule.RuleID)
+					}
+				}
+			}
+		}
+	}
+	return &ruleList
+}
+
+//Build policy ,rules and attached it to EPG
+func (k8sNet *k8sContext) applyContivNetworkPolicy(np *k8sNetworkPolicy) error {
+	var err error
+	// reset policy to deny on any error
+	policyResetOnErr := func(tenantName, groupName string) {
+		if err != nil {
+			//k8sNet.resetPolicy(tenantName, groupName)
+			npLog.Warnf("Need to reset the policy")
+		}
+	}
+	//endPoint Group Lookup(EPG)
+	if _, ok := k8sNet.epgName[np.PodSelector.GroupName]; !ok {
+		//Create EPG then
+		npLog.Infof("EPG :%v doesn't exist create now!",
+			np.PodSelector.GroupName)
+		if err = k8sNet.createEpgInstance(
+			np.PodSelector.NetworkName,
+			np.PodSelector.GroupName); err != nil {
+			npLog.Errorf("failed to create EPG %s ", err)
+			return err
+		}
+	}
+	//Get PolicyMap using EPG
+	policyMap := k8sNet.policyPerEpg[np.PodSelector.GroupName]
+	//Check if Policy is already programmed in EPG or not
+	if _, ok := k8sNet.networkPolicy[np.PodSelector.PolicyName]; !ok {
```

Comment:
policy names I think can be repeated across namespaces, this datastructure would have false positives for existence